Add initial version of navigation skill

primesun opened the pull request:
See #98

Stypox commented on app/src/main/java/org/dicio/dicio_android/skills/navigate/NavigateInfo.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Override
+    public boolean isAvailable(final SkillContext context) {
+        return true;
```
```suggestion
        return Sections.isSectionAvailable(navigate);
```

Stypox commented on app/src/main/sentences/en/navigate.dslf:
```diff
@@ -0,0 +1,2 @@
+navigate: high
+(navigate to?)|(take me to)|(directions to?)|(where s|is) .where.;
```
A few more ways to say things
```suggestion
(give? me? directions|direction|navigate to?)|(take|bring me to)|(where s|is) .where.;
```

primesun replied:
Wouldn't this also accept a command of "give navigate <address>"? Or "give me navigate to <address>"? Seems undesirable.

primesun replied:
I went ahead and added a modified version of your suggestion. Will resolve this conversation, but feel free to re-open.

Stypox commented on app/src/main/java/org/dicio/dicio_android/skills/navigate/NavigateProcessor.java:
```diff
@@ -0,0 +1,79 @@
+        final List<Object> textWithNumbers = ctx().requireNumberParserFormatter()
+                .extractNumbers(data.getCapturingGroup("where")).get();
+
+        // It seems that we detect each digit of the numerical portion of an address as a separate
+        // string, resulting in "9 5 4 6 1 9 avenue" when the speech was actually "9546 19 avenue".
+        // Here we will remove the spaces from between the numbers, so we will end up with
+        // "954619 avenue". This is not ideal, but it seems slightly better than before.
+        // Can we detect addresses better?
```
I am not so sure what user input you are referring to here. Wouldn't the user say "nine five four six nineteenth avenue" in that case? So the 19th is separate from the 9546, simply because it is ordinal.

primesun replied:
Without the number parser formatter, a speech input of "nine" would be transcribed as "nine", whereas we almost always would want that to be transcribed as "9". So that part certainly seems needed. With that out of the way:

For an address of "9546 19 avenue":
- If the speech input is "nine thousand five hundred forty six nineteen avenue" the transcribed output is **correct** as expected "9546 19 avenue"
- If the speech input is "nine thousand five hundred forty six nineteenth avenue" the transcribed output is **wrong** because it seems that the 9546 is multiplied by 1/19 resulting in an output of "502.421052... avenue"
- If the speech input is "nine five four six nineteen avenue" the transcribed output is **wrong** (but close): "9 5 4 6 19 avenue"
- If the speech input is "nine five four six nineteenth avenue" the transcribed output is **wrong**: "9 5 4 0.3157894... avenue". Here we have 6 multiplied by 1/19.
- If the speech input is "ninety five forty six nineteen avenue" the transcribed output is **wrong**: "95 46 19 avenue"

In summary, the problems are:
- Specifying the street number as an ordinal seems to result in some undesired multiplication happening
- Saying the building number using its digits one by one results in undesired spaces in between each one
- Saying the building number using partial grouping of its digits (but not all of them) e.g. "ninety five forty six" also results in undesired spaces in between each partial grouping

I have not attempted to resolve the ordinal multiplication problem at all. I suspect that might require some changes in the number formatter (maybe by allowing the caller to specify whether or not to perform calculations?)

In my code I had tried to resolve the latter 2 problems by removing the spaces in between any numerical strings, however this resulted in also removing the space in between the building number and the street number. For now, I propose removing that space-removal code, meaning we'll still have the original problems but it will work as expected if we specify each of building number and street number as full separate numbers rather than by digit.

primesun replied:
I proceeded with my proposal of removing the space-removal code. I documented the known issues in a comment beside the relevant code as well. Resolving this conversation, but feel free to re-open.

Stypox commented on app/src/main/java/org/dicio/dicio_android/skills/navigate/NavigateProcessor.java:
```diff
@@ -0,0 +1,79 @@
+        // Here we will remove the spaces from between the numbers, so we will end up with
+        // "954619 avenue". This is not ideal, but it seems slightly better than before.
+        // Can we detect addresses better?
+        placeToNavigate = "";
```
It might be better to user `StringBuilder`

Stypox commented on app/src/main/java/org/dicio/dicio_android/skills/navigate/NavigateProcessor.java:
```diff
@@ -0,0 +1,79 @@
+        while (i < textWithNumbers.size()) {
+            final Object currentItem = textWithNumbers.get(i);
+            if (!currentItem.toString().equals(" ")) {
+                if (i == 0 || isInteger(currentItem.toString())) {
+                    placeToNavigate += currentItem.toString().trim();
```
I don't think you need `isInteger` and `toString` here. `extractNumbers` already takes care of extracting numbers (class `org.dicio.numbers.util.Number`) and separating them from `String`s (the usual `java.lang.String`). You may want to take a look at `CalculatorProcessor` for reference. But I guess `dicio-numbers` would benefit from a bit more documentation ;-)

primesun replied:
With the space removal code (that I described in a separate comment) I do need them, because I need to know for each string if it contains a number or not, so I can decide whether or not to remove spaces between it and adjacent numbers.

In my other comment I proposed removing the space-removal code since it isn't perfect anyway. And if I do that, then I'll be removing this as well.

primesun replied:
This has been resolved due to the other changes that I mentioned.

Stypox commented on app/src/main/java/org/dicio/dicio_android/skills/navigate/NavigateOutput.java:
```diff
@@ -0,0 +1,45 @@
+    @Override
+    public void generate(final Data data) {
+        if (data.failed) {
+            final String message = "Specify where you want to navigate to";
```
You should not hardcode strings to show to the user here, otherwise they can't be localized. Please move them to `strings.xml`

Stypox commented on app/src/main/java/org/dicio/dicio_android/skills/navigate/NavigateOutput.java:
```diff
@@ -0,0 +1,45 @@
+                ctx().android(), message));
+
+        final String uriGeoSimple = String.format(Locale.ENGLISH,
+                "geo:0,0?q=%s", data.address);
```
This seems to work for some apps, although for StreetComplete it seems like the `0,0` makes the app go to the coordinates `0,0` instead of going to the query given at `?q=`. Is the `0,0` required for geo intents? Could you give some documentation as to where you took this from?

primesun replied:
`geo:0,0?q=my+street+address` is the generic Android intent to show an address on the map, as far as I can tell. I've never used StreetComplete, so I don't know why it doesn't support it.

There are also other, app-specific intents.

Links:
- https://developer.android.com/guide/components/intents-common#ViewMap
- https://osmand.net/docs/technical/algorithms/osmand-intents/

primesun replied:
For now I will resolve this as I have provided answers, don't see any further comments, and to facilitate completion of this PR before it gets too stale. Feel free to re-open.

Stypox reviewed (changes requested):
Thank you! Code is almost fine, just a few things to fix here and there. I tested and it already seems to work pretty well. In the future we could also add things like "navigate from A to B". Also, saying "where is A" should just search for "A" in the navigation app, while "navigate to A" should start a journey from the user's current position to A in the navigation app. I don't know if this is doable with Android geo intents, and feel free not to implement these things already in this PR.

Stypox reviewed (approved):
I rebased and now all of your commits appear as though they were mine. But your name+email correctly appears in the commit metadata, and I also added "Co-authored-by: primesun <primesun@example.com>" to commit messages but that didn't solve the issue either...

Anyway, I pushed a few commits to refine some things. "Disabling" fractions in the number parser can be done with `.preferOrdinal(true)`. I also used `String` instead of `NavigationOutput.Data` for simplicity. Finally, I made it so that output is requested again to the user in case the place he mentioned could not be understood the first time.

Thank you for this nice skill and for the research about navigation apps!

primesun commented:
> I rebased and now all of your commits appear as though they were mine. But your name+email correctly appears in the commit metadata, and I also added "Co-authored-by: primesun <[primesun@example.com](mailto:primesun@example.com)>" to commit messages but that didn't solve the issue either...

No problem, this is one reason why I never use git rebase ;)
